source/common/router/router.cc

```diff
@@ -780,5 +765,37 @@ ProdFilter::createRetryState(const RetryPolicy& policy, Http::HeaderMap& request
                                 priority);
 }
 
+void Filter::UpstreamRequest::setRequestEncoder(Http::StreamEncoder& request_encoder) {
+  request_encoder_ = &request_encoder;
+  // Now that there is an encoder, have the connection manager inform the manager when the
+  // downstream buffers are overrun.  This may result in immediate watermark callbacks referencing
+  // the encoder.
+  parent_.callbacks_->addDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+}
+
+void Filter::UpstreamRequest::clearRequestEncoder() {
+  // Before clearing the encoder, unsubscribe from callbacks.  This may result
+  // in some clean up calls to onBelowWriteBufferLowWatermark which reference
+  // the encoder.
+  if (request_encoder_) {
+    parent_.callbacks_->removeDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+  }
+  request_encoder_ = nullptr;
+}
+
+void Filter::UpstreamRequest::DownstreamWatermarkManager::onAboveWriteBufferHighWatermark() {
+  ASSERT(parent_.request_encoder_);
+  // The downstream connection is overrun.  Pause reads from upstream.
+  parent_.parent_.cluster_->stats().upstream_flow_control_paused_reading_total_.inc();
+  parent_.request_encoder_->getStream().readDisable(true);
+}
+
+void Filter::UpstreamRequest::DownstreamWatermarkManager::onBelowWriteBufferLowWatermark() {
+  ASSERT(parent_.request_encoder_);
+  // The downstream connection has buffer available.  Resume reads from upstream.
+  parent_.parent_.cluster_->stats().upstream_flow_control_resumed_reading_total_.inc();
+  parent_.request_encoder_->getStream().readDisable(false);
```

Comment:
> The codec test has this thing where all data is flushed immediately

FYI it doesn't have to be this way I don't think (I'm not exactly sure which flushing you are talking about), but it might be too much work to setup the mocks to do what you want.

> What I did was check out the old code where we readEnable on reset, stuck a boolean guard set true during runResetCallbacks and verified my integration test occasionally consumes data under the stack of runResetCallbacks (it does) and with rc == 0 (it does). That doesn't mean every type of disconnect and reset would work, but it does show that that particular one does.

OK. So which version do you prefer? The current code? Or what you had before? I'm fine either way, just want to know what to review. :)